src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -112,9 +119,13 @@ corec client_program where
           client_program \<psi> \<kappa> (C @ [i]) \<phi> |
         Cont (RollBackward q) \<Rightarrow>
           client_program \<psi> \<kappa> (roll_back \<psi> C q) \<phi> |
-        Cont AwaitReply \<Rightarrow> \<comment> \<open>only for this initial implementation\<close>
-          \<up> Done;
-          \<bottom>
+        Cont AwaitReply \<Rightarrow>
+          \<down> M. (partial_case M of
+            Cont (RollForward i) \<Rightarrow>
+              client_program \<psi> \<kappa> (C @ [i]) \<phi> |
+            Cont (RollBackward q) \<Rightarrow>
+              client_program \<psi> \<kappa> (roll_back \<psi> C q) \<phi>
+          )
       )
   )"
```

Comment:
It seems that you have only lifted the first restriction mentioned in the issue description, which is as follows:

> The server sticks to a single chain throughout protocol execution, which it receives upon startup as a program argument.

However, there is also the second restriction to lift:

> The client maintains its current chain only internally as a program argument “updated” upon recursive calls.

The client program has to communicate its current chain to its environment somehow.

Reply:
That's true. I'm not sure exactly when the client should notify its environment about its current chain because the client's chain update is a continuous process. However, the environment indeed needs the client's current chain at some point since it has to perform chain selection. Perhaps we should look at the real implementation and/or discuss this with Duncan. What do you think?

Reply:
I always assumed that the client program notifies the other parts of the client about each and every update of its candidate chain. After all, it is beneficial for these other parts to know about new information quickly, for example, in order to forward new blocks quickly to other peers or detect illegal data or the inappropriateness of the chain early. We can check with @dcoutts. I can ask him tomorrow; after all, he’s currently only a few meters away from me. :slightly_smiling_face:

Reply:
Actually, I guess that in the Haskell implementation the client program doesn’t maintain its current candidate chain in some local pure variable but rather directly in the mutable state that it shares with the other parts of the client (which allows for notifications of those other parts, since this mutable state uses STM). This would mean that notification about new candidate chain data is immediate.

Reply:
My guess was confirmed by @dcoutts: The client program in the Haskell implementation directly updates the mutable state that it shares with its local environment; so notifications about updates aren’t deliberately delayed. That said, the client program in the Haskell implementation also performs block validation and ends the protocol execution if it detects an invalid block. However, @dcoutts said that we should model only the actual chain synchronization, not any validation. As a result, our client program should just send updated chains to the environment immediately.

Reply:
Fixed in 9c0b607d5e72186839b9d5738ccf82d493236a6d, please check.

Reply:
I’ve made some changes:
* 9c7382d510fbd0a208346deb8a8704cadfd6b634
* 10011352f70cfb60daa234e02683ef9f9beb3835
* 64d3db9ae5f31f8e4853878fdfcfc3c50d368b9c

Reply:
Now that `client_chain_updates` has been renamed to `client_chains` it might make sense to change the name of the argument `u` (perhaps to `c`?)

Reply:
See my comment [above](https://github.com/input-output-hk/ouroboros-high-assurance/pull/93#discussion_r1407937635).

Reply:
And see my comment [above](https://github.com/input-output-hk/ouroboros-high-assurance/pull/93#discussion_r1407959443). :slightly_smiling_face: